mirgecom/artificial_viscosity.py

```diff
@@ -0,0 +1,329 @@
+    )
+
+    # RHS_av = div(R) volume part
+    div_r_vol = discr.weak_div(r)
+    # Total flux of grad(Q) across element boundaries
+    r_bnd_flux = (_facial_flux_r(discr, r_tpair=interior_trace_pair(discr, r))
+                  + sum(_facial_flux_r(discr, r_tpair=pb_tpair)
+                    for pb_tpair in cross_rank_trace_pairs(discr, r))
+                  + sum(bnd.av_flux(discr, btag, diffusion=r, **boundary_kwargs)
+                        for btag, bnd in boundaries.items()))
+
+    # Return the AV RHS term
+    return discr.inverse_mass(-div_r_vol + discr.face_mass(r_bnd_flux))
+
+
+def artificial_viscosity(discr, t, eos, boundaries, q, alpha, **kwargs):
+    """Interface :function:`av_operator` with backwards-compatible API."""
+    from warnings import warn
+    warn("Do not call artificial_viscosity; it is now called av_operator. This"
+         "function will disappear in 2021", DeprecationWarning, stacklevel=2)
+    return av_operator(discr=discr, boundaries=boundaries,
+        boundary_kwargs={"t": t, "eos": eos}, q=q, alpha=alpha, **kwargs)
+
+
+def smoothness_indicator(discr, u, kappa=1.0, s0=-6.0):
+    r"""Calculate the smoothness indicator.
+
+    Parameters
+    ----------
+    u: meshmode.dof_array.DOFArray
+        The field that is used to calculate the smoothness indicator.
+
+    kappa
+        An optional argument that controls the width of the transition from 0 to 1.
+    s0
+        An optional argument that sets the smoothness level to limit
+        on. Values in the range $(-\infty,0]$ are allowed, where $-\infty$ results in
+        all cells being tagged and 0 results in none.
+
+    Returns
+    -------
+    meshmode.dof_array.DOFArray
+        The elementwise constant values between 0 and 1 which indicate the smoothness
+        of a given element.
+    """
+    if not isinstance(u, DOFArray):
+        raise ValueError("u argument must be a DOFArray.")
+
+    actx = u.array_context
+
+    @memoize_in(actx, (smoothness_indicator, "smooth_comp_knl"))
+    def indicator_prg():
+        """Compute the smoothness indicator for all elements."""
+        from arraycontext import make_loopy_program
+        return make_loopy_program([
+            "{[iel]: 0 <= iel < nelements}",
+            "{[idof]: 0 <= idof < ndiscr_nodes_in}",
+            "{[jdof]: 0 <= jdof < ndiscr_nodes_in}",
+            "{[kdof]: 0 <= kdof < ndiscr_nodes_in}"
+            ],
+            """
+                result[iel,idof] = sum(kdof, vec[iel, kdof]               \
+                                             * vec[iel, kdof]             \
+                                             * modes_active_flag[kdof]) / \
+                                   sum(jdof, vec[iel, jdof]               \
+                                             * vec[iel, jdof]             \
+                                             + 1.0e-12 / ndiscr_nodes_in)
+            """,
+            name="smooth_comp",
+        )
```

Comment:
I think support for transforming (parallelizing) kernels via the loop induction variable names is going away. These should probably tag the inames similar to https://github.com/inducer/meshmode/blob/91f7e58dd336d067fc1202d016a5baccf8502f58/meshmode/discretization/connection/direct.py#L147-L149.

Reply:
Patch for this would be:
```diff
diff --git a/mirgecom/artificial_viscosity.py b/mirgecom/artificial_viscosity.py
index 33093b7..4f58a8c 100644
--- a/mirgecom/artificial_viscosity.py
+++ b/mirgecom/artificial_viscosity.py
@@ -272,7 +272,10 @@ def smoothness_indicator(discr, u, kappa=1.0, s0=-6.0):
     def indicator_prg():
         """Compute the smoothness indicator for all elements."""
         from arraycontext import make_loopy_program
-        return make_loopy_program([
+        from meshmode.transform_metadata import (ConcurrentElementInameTag,
+                                                 ConcurrentDOFInameTag)
+        import loopy as lp
+        t_unit = make_loopy_program([
             "{[iel]: 0 <= iel < nelements}",
             "{[idof]: 0 <= idof < ndiscr_nodes_in}",
             "{[jdof]: 0 <= jdof < ndiscr_nodes_in}",
@@ -288,6 +291,8 @@ def smoothness_indicator(discr, u, kappa=1.0, s0=-6.0):
             """,
             name="smooth_comp",
         )
+        return lp.tag_inames(t_unit, {"iel": ConcurrentElementInameTag(),
+                                      "idof": ConcurrentDOFInameTag()})
 
     @keyed_memoize_in(actx, (smoothness_indicator,
                              "highest_mode"),

```

Reply:
Thanks, @kaushikcfd! Should this change be made right away?

Reply:
If so, any chance I could get you to file a PR against the add-av-to-ns-actx PR #373?

Reply:
> Should this change be made right away?

The name-matching interface would be removed in '22, so not very high-priority for existing array context implementations. But for newer array-contexts, we aren't supporting name-based parallelization to begin with. 

> file a PR against the add-av-to-ns-actx

Sure! Moving the discussion to https://github.com/illinois-ceesd/mirgecom/pull/419. (Resolving here.)